### ISSUE TITLE
Mouse move may not be compatible with Python 3.7 and forward

### ISSUE THREAD

**Issue by McSinyx** [McSinyx/brutalmaze#10]
`maze.set_step` uses a hack on `dict` to perform O(n) Dijkstra. Prior to Python 3.7, `dict`'s key are sorted by their hashes, thus it's a O(1) insertion sorted data type for integers (whose range is arch-dependent). Since version 3.7, as [dictionary order is guaranteed to be insertion order](https://docs.python.org/3/whatsnew/3.7.html), the previously mentioned method might cause unexpected behaviours. More testing needs to be done.

**Comment by nghiaxlee** [comment]
Just curious, can you point out for me why your Dijkstra is linear?

**Comment by McSinyx** [comment]
I think the correct term would by (water) flooding algo. It's linear (complexity of size of the visible part of the map) long as insertion and deletion is O(1).

BTW I've just proof read my own code and realized that [the dict iterated by key](https://github.com/McSinyx/brutalmaze/blob/1e7e981e815dc848a55c1325f6a891b427eb75de/brutalmaze/maze.py#L452), not as an iterable, so the method is unaffected by the new change in py3.7.  Basically I was super sleepy when I filed this issue so when I read the py3.7 change log I was really panicked lmao. The comment on L451 is there maybe because I forgot to remove it after a bunch of trials on data types to optimize the method (if we follow the textbook and use `heapq`, the algo would be O(n.logn)). Now `defaultdict` is only used as a syntactic sugar to avoid manually inserting the next step (see L466). Using simply a list instead would work, but that would cost 3 more lines for range checking lol.

**Comment by McSinyx** [comment]
> Using simply a list instead would work, but that would cost 3 more lines for range checking lol.

[That definitely worth it](https://github.com/McSinyx/brutalmaze/commit/ef70806a48a6917f64110716ec08d9174626079a#diff-1021dbaa6dddb5697e5d95e4d5c3fbbbL486-R491), given the list initialization is more idiomatic and list random access is (theoretically) faster.  I know, micro-optimization is pointless, but it makes my conscious feels better.

BTW happy birthday :birthday:

**Comment by nghiaxlee** [comment]
Anw, thanks so much. Would not expect someone to say Hpbd thru GitHub tho =))

Quick question regarding your algorithm (maybe I'm just stupid, enlighten me then xD), why do you need `distance` variable btw? AFAIK, this is a FIFO queue, so actually the points which have smaller distance will always be popped out before those with greater distance, thus a nested list is redundant? If what I said is not true, then we might have problem with "shortest". If you want to keep `distance` for some reasons, then I suggest tying it with the tuple (`queue = [(self.destx, self.desty, 0)]`)

**Comment by McSinyx** [comment]
> AFAIK, this is a FIFO queue, so actually the points which have smaller distance will always be popped out before those with greater distance, thus a nested list is redundant?

You are absolutely correct!  I guess [when I first implemented this](https://github.com/McSinyx/brutalmaze/commit/9dff378b57fbe16aa9edde9e3bc4eab959081e83), I was blindly focusing on how to simplify the use of priority queue and missed the incremental property of forest fire.  Once it is implemented, I haven't got any second thought about it ever since, i.e. don't fix what ain't broken.  This repo contains a lot of that kind of *legacy code* that I'm trying to fix, mainly for it to run well on phosh (Librem 5's interface).

Thank you for pointing this out, I've improved the snippet in b70c00e.
